jlm/util/GraphWriter.hpp

```diff
@@ -0,0 +1,843 @@
+  void
+  SetHtmlTableAttribute(std::string name, std::string value);
+
+  void
+  SetFillColor(std::string color) override;
+
+  void
+  Finalize() override;
+
+  void
+  OutputASCII(std::ostream & out, size_t indent) override;
+
+  void
+  OutputDot(std::ostream & out, size_t indent) override;
+
+  void
+  OutputSubgraphs(std::ostream & out, GraphOutputFormat format, size_t indent) override;
+
+private:
+  // Attributes that need to be placed on the HTML table in the dot output, and not on the node.
+  std::unordered_map<std::string, std::string> HtmlTableAttributes_;
+
+  std::vector<std::unique_ptr<InputPort>> InputPorts_;
+  std::vector<std::unique_ptr<OutputPort>> OutputPorts_;
+  std::vector<Graph *> SubGraphs_;
+};
+
+/**
+ * Node representing a single sink port
+ */
+class ArgumentNode : public Node
+{
+  friend Graph;
+
+  ArgumentNode(Graph & graph);
+
+public:
+  ~ArgumentNode() override = default;
+
+  const char *
+  GetIdPrefix() const override;
+
+  bool
+  CanBeEdgeHead() const override;
+
+  /**
+   * Indicate that the argument node represents a value coming in from a port in another graph
+   * @param outsideSource the Port in the other graph
+   */
+  void
+  SetOutsideSource(Port & outsideSource);
+
+  void
+  OutputASCII(std::ostream & out, size_t indent) override;
+
+private:
+  // Optional reference to a Port outside of this graph from which this argument came
+  Port * OutsideSource_;
+};
+
+/**
+ * Node representing a single source port
+ */
+class ResultNode : public Node
+{
+  friend Graph;
+
+  ResultNode(Graph & graph);
+
+public:
+  ~ResultNode() override = default;
+
+  const char *
+  GetIdPrefix() const override;
+
+  bool
+  CanBeEdgeTail() const override;
+
+  /**
+   * Indicate that the result node represents the value of a port in another graph
+   * @param outsideSource the Port in the other graph
+   */
+  void
+  SetOutsideDestination(Port & outsideSource);
+
+  void
+  OutputASCII(std::ostream & out, size_t indent) override;
+
+private:
+  // Optional reference to a Port outside of this graph representing where the result ends up
+  Port * OutsideDestination_;
+};
+
+class Edge : public GraphElement
+{
+  friend Graph;
+
+  Edge(Port & from, Port & to, bool directed);
+
+public:
+  ~Edge() override = default;
+
+  const char *
+  GetIdPrefix() const override;
+
+  Graph &
+  GetGraph() override;
+
+  /**
+   * Gets the port being pointed to
+   * Even if the edge is non-directed, the from/to order can matter for layout.
+   */
+  [[nodiscard]] Port &
+  GetFrom();
+
+  /**
+   * Gets the port being pointed from.
+   * Even if the edge is non-directed, the from/to order can matter for layout.
+   */
+  [[nodiscard]] Port &
+  GetTo();
+
+  /**
+   * @return true if this edge is directed, false otherwise
+   */
+  [[nodiscard]] bool
+  IsDirected() const;
+
+  /**
+   * Given one end of the edge, returns the port on the opposite side of the edge.
+   */
+  [[nodiscard]] Port &
+  GetOtherEnd(Port & from);
```

Comment:
The variable name `from` throws me off here. I expected this method to only return the `to` port, but given either of them, it returns the other.